pkg/machine/machineset.go

```diff
@@ -0,0 +1,538 @@
+	}
+
+	glog.V(100).Infof("Setting GCPMachineProviderConfig.MachineType param value to: %s", machineType)
+
+	GCPProviderSpecObject.MachineType = machineType
+
+	byteArrayGCP, err := json.Marshal(GCPProviderSpecObject)
+
+	if err != nil {
+		glog.V(100).Infof("error marshalling GCPMachineProviderSpec object into byte array: %v", err)
+
+		return fmt.Errorf("could not update MachineType param: %w", err)
+	}
+
+	err = json.Unmarshal(byteArrayGCP, builder.Definition.Spec.Template.Spec.ProviderSpec.Value)
+
+	if err != nil {
+		glog.V(100).Infof("error unmarshalling ProviderSpec byte array into ProviderSpec.Value: %v", err)
+
+		return fmt.Errorf("could not update MachineType param: %w", err)
+	}
+
+	return nil
+}
+
+// AzureChangeProviderVMSize changes the ProviderSpec VMSize param for Azure public cloud.
+func (builder *MachineSetBuilder) AzureChangeProviderVMSize(byteArray []byte, vmSize string) error {
+	if valid, err := builder.validate(); !valid {
+		return err
+	}
+
+	if vmSize == "" {
+		return fmt.Errorf("vmSize parameter cannot be empty")
+	}
+
+	glog.V(100).Infof("Updating ProviderSpec Value VMSize param '%s' for Azure public cloud",
+		vmSize)
+
+	var AzureProviderSpecObject *machinev1beta1.AzureMachineProviderSpec
+	err := json.Unmarshal(byteArray, &AzureProviderSpecObject)
+
+	if err != nil {
+		glog.V(100).Infof("error unmarshalling byte array into AzureMachineProviderSpec object: %v", err)
+
+		return fmt.Errorf("could not update VMSize param: %w", err)
+	}
+
+	glog.V(100).Infof("Setting AzureMachineProviderSpec.VMSize param value to: %s", vmSize)
+
+	AzureProviderSpecObject.VMSize = vmSize
+
+	byteArrayAzure, err := json.Marshal(AzureProviderSpecObject)
+
+	if err != nil {
+		glog.V(100).Infof("error marshalling AzureMachineProviderSpec object into byte array: %v", err)
+
+		return fmt.Errorf("could not update VMSize param: %w", err)
+	}
+
+	err = json.Unmarshal(byteArrayAzure, builder.Definition.Spec.Template.Spec.ProviderSpec.Value)
+
+	if err != nil {
+		glog.V(100).Infof("error unmarshalling AzureMachineProviderSpec byte array into "+
+			"ProviderSpec.Value object: %v", err)
+
+		return fmt.Errorf("could not update VMSize param: %w", err)
+	}
+
+	return nil
+}
+
+// ChangeCloudProviderInstanceType calls the cloud-specific function to change the ProviderSpec instance type param.
+func (builder *MachineSetBuilder) ChangeCloudProviderInstanceType(instanceType string) error {
+	if valid, err := builder.validate(); !valid {
+		return err
+	}
+
+	glog.V(100).Infof("Updating the cloud provider instance type field")
+
+	byteArray, err := json.Marshal(builder.Definition.Spec.Template.Spec.ProviderSpec.Value)
+
+	if err != nil {
+		return fmt.Errorf("error marshalling machineSet providerSpec.Value into byte array: %w", err)
+	}
+
+	switch builder.publicCloud {
+	case "aws":
+		glog.V(100).Infof("Updating ProviderSpec InstanceType param for AWS public cloud")
+
+		err = builder.AWSChangeProviderInstanceType(byteArray, instanceType)
```

Comment:
Personally I don't see the need to pass in byteArray when it can be accessed via builder.Definition.Spec.Template.Spec.ProviderSpec.Value in the method

Reply:
Done